### ISSUE TITLE
Feature classes are not identical for model and x_train. model provided: ,

### ISSUE THREAD

**Issue by djd224** [NorskRegnesentral/shapr#340]
Hello all!

I am interested in importing my PyTorch model into R in order to use the shapr package for further analysis of the model.

I trained the model in PyTorch, implemented it in R using the relatively new 'torch' package, and am now following the custom model example from [this vignette](https://norskregnesentral.github.io/shapr/articles/understanding_shapr.html#compare) to apply shapr.  However, when running the explain function, I get the following error:

`Error in compare_vecs(spec1$classes[sort(names(spec1$classes))], spec2$classes[sort(names(spec2$classes))],  : 
  Feature classes are not identical for model and x_train.
model provided: ,
x_train provided: ASA_m^2/g:: , AV_cm^3/g:: , Density:: , GeDi: , LabuteASA: , lsd_f: , lsd_p: , MW: , PBF: , PMI1: , PMI2: , PMI3: , SPAN: , Vol: .`

My interpretation of this error is that the feature names in the dataset I feed the model don't match those in the model.  However, when I run my custom `MY_get_model_specs()`, the labels DO match those provided by x_train.  I am confused why it appears as though my model is not providing any feature classes.

Any guidance on this issue would be greatly appreciated. Thank you!

**Comment by martinju** [comment]
Hi!

That's odd. If you can provide a minimal working example with your custom functions, I can take a look.

**Comment by djd224** [comment]
I'm assuming things might be getting a little funky because of my use of torch as from my understanding it is a relatively new package.  Here are the custom functions I have written for my model:

`X_train <- data.table::as.data.table(X_train)
data.table::setnames(X_train, as.character(X_keys))

X_test <- data.table::as.data.table(X_test)
data.table::setnames(X_test, as.character(X_keys))


MY_predict_model <- function(x, newdata) {
  newdata <- torch_tensor(newdata)
  prediction <- as_array(x(newdata)[[1]]$detach())
  return(prediction)
}

MY_get_model_specs <- function(x) {
  feature_specs <- list()
  feature_specs$labels <- X_keys
  m <- length(feature_specs$labels)
  feature_specs$classes <- rep("numeric", m)
  feature_specs$factor_levels <- setNames(vector("list", m), feature_specs$labels)
  feature_specs$factor_levels[feature_specs$classes == "factor"] <- NA
  return(feature_specs)
  #return(list(labels = feature_specs$labels, classes = feature_specs$classes, factor_levels = feature_specs$factor_levels))
}

# explainer
set.seed(123)
p0 <- mean(y_train)

explanation_custom <- explain(
  model = model,
  x_explain = X_test,
  x_train = X_train,
  approach = "empirical",
  prediction_zero = p0,
  predict_model = MY_predict_model,
  get_model_specs = MY_get_model_specs
)`

`X_keys` contains the strings associated with each feature name, in accordance with those provided by x_train.  I hope this helps!

**Comment by martinju** [comment]
The point with "get_model_specs" is to extract the features from the model object (x). When you just provide them by passing the global variable X_keys, you may as well skip the entire function (it is not required, but will then show warnings). Have you tried that?

If you want to do it properly, enabling the extra checks, you need to find out how to extract the feature names from the torch model object itself.

If you still got problems, and want me to take a look, please provide a complete runnable example that produces the issue.

**Comment by djd224** [comment]
Thank you for your help, I will keep working on it and get back to you as soon as I can.
